api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+As an example, the following are the expansions of some instructions.
+
+Expansion for `vpgatherdd`: https://gist.github.com/abhinav92003/637c07bac785428550f05c18319f80c2
+
+Expansion for `vpscatterdd`: https://gist.github.com/abhinav92003/4a78ec4ad1621cbf6ca905a6774cb389
+
+We also need to make sure that if some load or store faults, we restore the application state properly. While the spilled registers obtained from drreg are restored by the drreg state restoration logic, drx still needs to restore the scratch mask registers that are spilled manually to a GPR and ensure that the bit for the faulting access is cleared too. In case a fault or some other state restore event like DR detach happens, we walk the expanded sequence using a state machine till we reach the fault pc, keeping track of the state that needs to be restored ([commit](https://github.com/DynamoRIO/dynamorio/commit/de511a72197e9b04a9732f33622fe7bc7fd12623), [commit](https://github.com/DynamoRIO/dynamorio/commit/e54913b4845794b96e8682ba6a214d54e10741a0)).
+
+
+## Drreg Support For Multi-phase Reservations
+
+Owner: [Abhinav Sharma](https://github.com/abhinav92003)
+
+Upstream issue: [DynamoRIO/dynamorio#3823](https://github.com/DynamoRIO/dynamorio/issues/3823)
+
+Drreg is DynamoRIO’s register reservation framework. It allows users to reserve a reg to use as scratch. Internally, drreg automatically performs the following functions so that the user does not need to. Drreg
+- keeps all required book-keeping like the spill slot to spilled register mapping
+- restores spilled registers to their application value before they are read by an application instruction; also, it re-spills the spilled registers if they are written by an application instruction.
+- performs application state restoration on state restore events like encountering an application fault, and DR detach.
+
+While expanding a scatter or gather instruction in the app2app phase, we need a scratch register to hold the scalar values and masks. In later phases (like the insertion or the instru2instru phase), drcachesim and other DR clients may also use drreg to get scratch registers for their instrumentation.
+
+Drreg initially supported only single phase use and did not keep any state between phases. To support multi-phase use of drreg, we need to solve the following:
```

Comment:
Make it more clear that the single phase supported was insertion phase (o/w it sounds like it could be any of them).  There was non-insertion-phase support (the key functions worked), there was just no explicit support for bad interactions between them; the non-insertion phase used was assumed to be instru2instru or sthg where the client had to be careful.

Reply:
Done